File: dotnet/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/ProtectionProxy.cs
namespace ProxyDesignPattern.Example1;

/// <summary>
/// PROTECTION PROXY - Controls access based on user subscription.
/// Free users cannot access premium content.
/// </summary>
public class ProtectionProxy : IVideoService
{
    private readonly IVideoService _innerService;
    private readonly User _currentUser;

    public ProtectionProxy(IVideoService innerService, User currentUser)
    {
        _innerService = innerService;
        _currentUser = currentUser;
    }

    public Video? GetVideoInfo(string videoId)
    {
        // Everyone can see video info
        return _innerService.GetVideoInfo(videoId);
    }

    public void StreamVideo(string videoId)
    {
        var video = _innerService.GetVideoInfo(videoId);
        
        if (video == null)
        {
            Console.WriteLine($"[ProtectionProxy] Video not found: {videoId}");
            return;
        }

        // Check if user has access to premium content
        if (video.IsPremium && _currentUser.Subscription == SubscriptionType.Free)
        {
            Console.WriteLine($"[ProtectionProxy] ACCESS DENIED: '{_currentUser.Name}' needs Premium subscription");
            Console.WriteLine($"[ProtectionProxy] Upgrade to Premium to watch: {video.Title}");
            return;
        }

        Console.WriteLine($"[ProtectionProxy] Access granted for user: {_currentUser.Name}");
        _innerService.StreamVideo(videoId);
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/ProxyDesignPattern/ProxyDesignPattern.Example1/src/ProtectionProxy.ts
import { IVideoService } from './IVideoService';
import { Video, User, SubscriptionType } from './Models';

// PROTECTION PROXY — controls access based on subscription
export class ProtectionProxy implements IVideoService {
  constructor(
    private readonly innerService: IVideoService,
    private readonly currentUser: User,
  ) {}

  getVideoInfo(videoId: string): Video | null {
    return this.innerService.getVideoInfo(videoId);
  }

  streamVideo(videoId: string): void {
    const video = this.innerService.getVideoInfo(videoId);
    if (!video) {
      console.log(`[ProtectionProxy] Video not found: ${videoId}`);
      return;
    }
    if (video.isPremium && this.currentUser.subscription === SubscriptionType.Free) {
      console.log(`[ProtectionProxy] ACCESS DENIED: '${this.currentUser.name}' needs Premium subscription`);
      console.log(`[ProtectionProxy] Upgrade to Premium to watch: ${video.title}`);
      return;
    }
    console.log(`[ProtectionProxy] Access granted for user: ${this.currentUser.name}`);
    this.innerService.streamVideo(videoId);
  }
}
